lib/libp2phttp.go

```diff
@@ -0,0 +1,156 @@
+package vole
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"fmt"
+	"math/big"
+	"net"
+	"net/http"
+	"net/http/httputil"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/peer"
+	libp2phttp "github.com/libp2p/go-libp2p/p2p/http"
+	"github.com/multiformats/go-multiaddr"
+)
+
+func Libp2pHTTPSocketProxy(ctx context.Context, p multiaddr.Multiaddr, unixSocketPath string) error {
+	h, err := libp2pHost()
+	if err != nil {
+		return err
+	}
+
+	httpHost := libp2phttp.Host{StreamHost: h}
+
+	ai := peer.AddrInfo{
+		Addrs: []multiaddr.Multiaddr{p},
+	}
+	idStr, err := p.ValueForProtocol(multiaddr.P_P2P)
+	if err == nil {
+		id, err := peer.Decode(idStr)
+		if err != nil {
+			return err
+		}
+		ai.ID = id
+	}
+
+	hasTLS := false
+	hasHTTP := false
+	multiaddr.ForEach(p, func(c multiaddr.Component) bool {
+		if c.Protocol().Code == multiaddr.P_HTTP {
+			hasHTTP = true
+		}
+
+		if c.Protocol().Code == multiaddr.P_HTTPS {
+			hasHTTP = true
+			hasTLS = true
+			return false
+		}
+
+		if c.Protocol().Code == multiaddr.P_TLS {
+			hasTLS = true
+		}
+		return true
+	})
+
+	rt, err := httpHost.NewConstrainedRoundTripper(ai)
+	if err != nil {
+		return err
+	}
+	rp := &httputil.ReverseProxy{
+		Transport: rt,
+		Director:  func(r *http.Request) {},
+	}
+
+	// Serves an HTTP server on the given path using unix sockets
+	server := &http.Server{
+		Handler: rp,
+	}
+
+	l, err := net.Listen("unix", unixSocketPath)
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		<-ctx.Done()
+		server.Close()
+	}()
+
+	if hasTLS && hasHTTP {
+		c, err := selfSignedTLSConfig()
+		if err != nil {
+
+			return err
+		}
+		server.TLSConfig = c
+
+		fmt.Println("Endpoint is an HTTPS endpoint. Using a self signed cert locally to proxy.")
```

Comment:
Maybe this is obvious, but can you give me some context on the value of "if the endpoint is HTTPS, let's add a self-signed cert on the unix socket proxy" rather not doing it at all or having the self-signed cert being a separate flag

Reply:
Imagine you run:
```
$ vole libp2p http --socket-path=/tmp/vole.sock /dns/google.com/https
```

What do you expect the curl request to look like?

I would expect:
```
curl --unix-socket /tmp/vole.sock https://www.google.com
```

Not 
```
curl --unix-socket /tmp/vole.sock http://www.google.com
```

`https://` vs `http://`

The only way we can do that is if our proxy itself is an https server, and thus we need a cert. Self signed is easiest. So we actually need to add a -k to the curl command.

Reply:
> What do you expect the curl request to look like?

I guess this is probably my issue. Because I'm having trouble seeing the value of the HTTP -> HTTP proxy other than to demonstrate that it's possible via libp2p multiaddrs the need to drop an `s` vs to add a `-k` doesn't mean a ton to me.

On the other hand it doesn't really hurt, so as long as it's not causing trouble and you think it's useful I don't really mind

Reply:
> Because I'm having trouble seeing the value of the HTTP -> HTTP proxy other than to demonstrate that it's possible via libp2p multiaddrs 

This is not the main use case, but it's nice that it works consistently. You can make the exact same curl request to a standard HTTP server as you would to a libp2p node over a stream.